api/v1/properties.go

```diff
@@ -47,26 +52,75 @@ type LinstorNodePropertyValueFrom struct {
+	// If set, the templates is appended to the defined property name, creating multiple properties instead of one
+	// aggregate.
+	// * $1 is replaced with the matched key.
+	// * $2 is replaced with the matched value.
+	//+kubebuilder:validation:Optional
+	NameTemplate string `json:"nameTemplate,omitempty"`
+
+	// ValueTemplate defines how the property value is expanded.
+	// * $1 is replaced with the matched key.
+	// * $2 is replaced with the matched value.
+	//+kubebuilder:validation:Optional
+	ValueTemplate string `json:"valueTemplate,omitempty"`
+
+	// Delimiter used to join multiple key and value pairs together.
+	//+kubebuilder:validation:Optional
+	Delimiter string `json:"delimiter,omitempty"`
+}
+
 func ValidateNodeProperties(props []LinstorNodeProperty, path *field.Path) field.ErrorList {
 	var result field.ErrorList
 
 	for i := range props {
 		p := &props[i]
 
-		valSet := p.Value != ""
-		fromSet := p.ValueFrom != nil
-		if valSet == fromSet {
-			result = append(result, field.Invalid(path.Child(strconv.Itoa(i)), p, "Expected exactly one of 'value' and 'valueFrom' to be set"))
+		sourcesSet := 0
+
+		if p.Value != "" {
+			sourcesSet++
+		}
+
+		if p.ValueFrom != nil {
+			sourcesSet++
 		}
 
-		if fromSet {
+		if p.ExpandFrom != nil {
+			sourcesSet++
+		}
+
+		if sourcesSet != 1 {
+			result = append(result, field.Invalid(path.Child(strconv.Itoa(i)), p, "Expected exactly one of 'value', 'valueFrom' or 'joinValuesFrom' to be set"))
+		}
+
+		if p.ValueFrom != nil {
 			_, keys, err := fieldpath.ExtractFieldPath(&corev1.Node{}, p.ValueFrom.NodeFieldRef)
 			if err != nil {
 				result = append(result, field.Invalid(path.Child(strconv.Itoa(i), "valueFrom", "nodeFieldRef"), p.ValueFrom.NodeFieldRef, fmt.Sprintf("Invalid reference format: %s", err)))
 			}
 
-			if keys != nil && !strings.Contains(p.Name, "$1") {
-				result = append(result, field.Invalid(path.Child(strconv.Itoa(i), "name"), p.Name, "Wildcard property requires replacement target `$1` in name"))
+			if keys != nil {
+				result = append(result, field.Invalid(path.Child(strconv.Itoa(i), "valueFrom", "nodeFieldRef"), p.ValueFrom.NodeFieldRef, "Wildcard property not allowed, use expandFrom instead"))
+			}
+		}
+
+		if p.ExpandFrom != nil {
+			_, keys, err := fieldpath.ExtractFieldPath(&corev1.Node{}, p.ExpandFrom.NodeFieldRef)
+			if err != nil {
+				result = append(result, field.Invalid(path.Child(strconv.Itoa(i), "expandFrom", "nodeFieldRef"), p.ExpandFrom.NodeFieldRef, fmt.Sprintf("Invalid reference format: %s", err)))
+			}
+
+			if keys == nil {
+				result = append(result, field.Invalid(path.Child(strconv.Itoa(i), "expandFrom", "nodeFieldRef"), p.ExpandFrom.NodeFieldRef, "Wildcard property required"))
+			}
+
+			if p.ExpandFrom.NameTemplate != "" && p.ExpandFrom.Delimiter != "" {
+				result = append(result, field.Invalid(path.Child(strconv.Itoa(i), "expandFrom"), p.ExpandFrom, "Expected exactly one of 'nameTemplate' and 'delimiter' to be set"))
```

Comment:
This only complains if both are set, rather than "exactly one" as the error message says. Do we need a check for neither being set? Or that allowed and we use a default delimiter?

Reply:
Adapted the error message. I don't know what use an empty `nameTemplate` and empty `delimiter` would be, but an empty delimiter may be desired in some cases, and I don't want to switch to using `*string` for "empty" values.